File: addons/terrabrush/Scripts/Tools/ObjectTool.cs
#if TOOLS
using System.Collections.Generic;
using Godot;

namespace TerraBrush;

public class ObjectTool : ToolBase {
    private HashSet<ZoneResource> _sculptedZones;
    private Dictionary<ZoneResource, Image> _heightmapImagesCache = null;
    private Dictionary<ZoneResource, Image> _waterImagesCache = null;

    public ObjectTool(TerraBrush terraBrush) : base(terraBrush) {}

    public override void BeginPaint() {
        base.BeginPaint();

        _heightmapImagesCache = new Dictionary<ZoneResource, Image>();
        _waterImagesCache = new Dictionary<ZoneResource, Image>();

        _sculptedZones = new HashSet<ZoneResource>();
    }

    public override void EndPaint() {
        base.EndPaint();

        _heightmapImagesCache = null;
        _waterImagesCache = null;

        _sculptedZones = null;
    }

    protected override ImageTexture GetToolCurrentImageTexture(ZoneResource zone) {
        return zone.ObjectsTexture[_terraBrush.ObjectIndex.Value];
    }

    public override void Paint(TerrainToolType toolType, Image brushImage, int brushSize, float brushStrength, Vector2 imagePosition) {
        if (_terraBrush.ObjectIndex == null) {
            return;
        }

        var currentObject = _terraBrush.Objects[_terraBrush.ObjectIndex.Value];
        if (currentObject?.Definition?.ObjectScenes == null || currentObject.Hide) {
            return;
        }

        var noiseTexture = currentObject.Definition?.NoiseTexture ?? _terraBrush.DefaultNoise;
        Image noiseImage = null;
        if (noiseTexture != null) {
            noiseImage = noiseTexture.GetImage();
        }

        var objectsNode = _terraBrush.ObjectsContainerNode.GetNode<IObjectsNode>($"{_terraBrush.ObjectIndex}");

        ForEachBrushPixel(brushImage, brushSize, imagePosition, (imageZoneInfo, pixelBrushStrength) => {
            _heightmapImagesCache.TryGetValue(imageZoneInfo.Zone, out var heightmapImage);
            if (heightmapImage == null) {
                heightmapImage = imageZoneInfo.Zone.HeightMapTexture.GetImage();
                _heightmapImagesCache.Add(imageZoneInfo.Zone, heightmapImage);
            }

            Image waterImage = null;
            if (_terraBrush.WaterDefinition != null) {
                _waterImagesCache.TryGetValue(imageZoneInfo.Zone, out waterImage);
                if (waterImage == null) {
                    waterImage = imageZoneInfo.Zone.WaterTexture.GetImage();
                    _waterImagesCache.Add(imageZoneInfo.Zone, waterImage);
                }
            }

            var xPosition = imageZoneInfo.ZoneInfo.ImagePosition.X;
            var yPosition = imageZoneInfo.ZoneInfo.ImagePosition.Y;

            var currentPixel = imageZoneInfo.Image.GetPixel(xPosition, yPosition);
            var newColor = currentPixel;

            if (pixelBrushStrength > 0f) {
                newColor = toolType == TerrainToolType.ObjectAdd ? Colors.Red : Colors.Transparent;

                objectsNode.AddRemoveObjectFromTool(toolType == TerrainToolType.ObjectAdd, xPosition, yPosition, imageZoneInfo.Zone, heightmapImage, waterImage, noiseImage);
                imageZoneInfo.Image.SetPixel(xPosition, yPosition, newColor);

                _sculptedZones.Add(imageZoneInfo.Zone);
            }
        });

        objectsNode.UpdateMeshesFromTool();
    }
}
#endif

File: src/editor_tools/object_tool.cpp
#include "object_tool.h"
#include "../nodes/objects_base.h"
#include "../editor_resources/object_resource.h"
#include "../editor_resources/object_definition_resource.h"

using namespace godot;

void ObjectTool::_bind_methods() {}

ObjectTool::ObjectTool() {}

ObjectTool::~ObjectTool() {}

Ref<Image> ObjectTool::getToolCurrentImage(Ref<ZoneResource> zone) {
    return zone->get_objectsImage()[_selectedObjectIndex];
}

void ObjectTool::beginPaint() {
    ToolBase::beginPaint();

    _sculptedZones = std::unordered_set<Ref<ZoneResource>>();
}

void ObjectTool::endPaint() {
    ToolBase::endPaint();

    _sculptedZones = std::unordered_set<Ref<ZoneResource>>();
}

void ObjectTool::paint(TerrainToolType toolType, Ref<Image> brushImage, int brushSize, float brushStrength, Vector2 imagePosition) {
    if (_selectedObjectIndex < 0) {
        return;
    }

    Ref<ObjectResource> currentObject = _terraBrush->get_objects()[_selectedObjectIndex];
    if (currentObject->get_definition()->get_objectScenes().size() == 0 || currentObject->get_hide()) {
        return;
    }

    ObjectsBase *objectsNode = Object::cast_to<ObjectsBase>(_terraBrush->get_objectsContainer()->get_node_or_null(String::num_int64(_selectedObjectIndex)));

    forEachBrushPixel(brushImage, brushSize, imagePosition, ([&](ImageZoneInfo &imageZoneInfo, float pixelBrushStrength) {
        Ref<Image> heightmapImage = imageZoneInfo.zone->get_heightMapImage();
        Ref<Image> waterImage = nullptr;
        if (!_terraBrush->get_waterDefinition().is_null()) {
            waterImage = imageZoneInfo.zone->get_waterImage();
        }

        int xPosition = imageZoneInfo.zoneInfo.imagePosition.x;
        int yPosition = imageZoneInfo.zoneInfo.imagePosition.y;

        Color currentPixel = imageZoneInfo.image->get_pixel(xPosition, yPosition);
        Color newColor = currentPixel;

        if (pixelBrushStrength > 0) {
            newColor = toolType == TerrainToolType::TERRAINTOOLTYPE_OBJECTADD ? Color(1, 0, 0, 1) : Color(1, 1, 1, 0);

            objectsNode->addRemoveObjectFromTool(toolType == TerrainToolType::TERRAINTOOLTYPE_OBJECTADD, xPosition, yPosition, imageZoneInfo.zone, heightmapImage, waterImage);
            imageZoneInfo.image->set_pixel(xPosition, yPosition, newColor);

            _sculptedZones.insert(imageZoneInfo.zone);
        }
    }));

    objectsNode->updateMeshesFromTool();
}

void ObjectTool::updateSelectedObjectIndex(int value) {
    _selectedObjectIndex = value;
}